### PR TITLE
support for setting labels to existing nodes

### PR DESCRIPTION
**Pull request by DvirDukhan**
WIP

This PR allows adding labels to existing nodes by calling:
`MATCH (n) SET n:Label1` 
`MATCH(n) SET n:Label1:Label2`
`MATCH(n) SET n:Label1, n:Label2`
`MATCH(n) SET n:Label1, n.prop="value"`
`MATCH(n), (m) SET n:Label1, n.prop="value", m:Label2`

- [x] support adding  update labels op to execution plan
- [x] flow tests
- [x] TCK

### CODE REVIEW

File: src/execution_plan/execution_plan_build/execution_plan_construct.c [hunk 1]
```diff
@@ -154,9 +154,17 @@ static inline void _buildUnwindOp(ExecutionPlan *plan, const cypher_astnode_t *c
...
+	rax *label_updates = raxNew();
+	AST_PrepareUpdateOp(gc, clause, property_updates, label_updates);
+	if(raxSize(property_updates) != 0) {
+		OpBase *op = NewUpdateOp(plan, property_updates);
+		ExecutionPlan_UpdateRoot(plan, op);
+	}
+	if(raxSize(label_updates) != 0) {
+		OpBase *op = NewUpdateLabelsOp(plan, label_updates);
+		ExecutionPlan_UpdateRoot(plan, op);
+	}
```
Comment (AviAvni) [inline src/execution_plan/execution_plan_build/execution_plan_construct.c:167]: if you don't create the operation you need to free the max

File: src/execution_plan/ops/op_update_labels.c [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+	array_free(op->node_updates);
+	op->node_updates = NULL;
+
+	op->updates_committed = false;
+	return OP_OK;
+}
+
+
+static void UpdateLabelsFree(OpBase *ctx) {
+	OpUpdateLabels *op = (OpUpdateLabels *)ctx;
```
Comment (AviAvni) [inline src/execution_plan/ops/op_update_labels.c:126]: fix indentation

File: src/ast/ast_build_op_contexts.c [hunk 3]
```diff
@@ -98,29 +99,51 @@ static void _ConvertSetItem(GraphContext *gc, rax *updates,
...
+				UpdateCtx_Clear(ctx);
+				UpdateCtx_SetMode(ctx, UPDATE_REPLACE);
+			}
 		}
-	}
 
-	// updated value
-	AR_ExpNode *exp = AR_EXP_FromASTNode(ast_value);
+		// updated value
+		AR_ExpNode *exp = AR_EXP_FromASTNode(ast_value);
```
Comment (AviAvni) [inline src/ast/ast_build_op_contexts.c:142]: indentation

File: src/ast/ast_shared.h [hunk 4]
```diff
@@ -54,6 +54,7 @@ typedef enum {
 	UPDATE_UNSET   = 0,    // default, should not be encountered
 	UPDATE_MERGE   = 1,    // merge new properties into existing property map
 	UPDATE_REPLACE = 2,    // replace existing property map with new properties
+	SET_LABELS 	   = 3,		   // add labels to the entity.
```
Comment (AviAvni) [inline src/ast/ast_shared.h:57]: fix space

File: src/execution_plan/ops/op_update_labels.c [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+static Record UpdateLabelsConsume(OpBase *opBase);
+static OpResult UpdateLabelsReset(OpBase *opBase);
+static OpBase *UpdateLabelsClone(const ExecutionPlan *plan, const OpBase *opBase);
+static void UpdateLabelsFree(OpBase *opBase);
+
+static Record _handoff(OpUpdateLabels *op) {
+	/* TODO: poping a record out of op->records
+	 * will reverse the order in which records
+	 * are passed down the execution plan. */
+	if(op->records && array_len(op->records) > 0) return array_pop(op->records);
```
Comment (AviAvni) [inline src/execution_plan/ops/op_update_labels.c:16]: is this todo need to be fixed?

File: src/execution_plan/ops/op_update_labels.c [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+
+	// done reading; we're not going to call Consume any longer
+	// there might be operations like "Index Scan" that need to free the
+	// index R/W lock - as such, free all ExecutionPlan operations up the chain.
+	OpBase_PropagateFree(child);
+
+	// lock everything
+	QueryCtx_LockForCommit();
+	{
+
```
Comment (AviAvni) [inline src/execution_plan/ops/op_update_labels.c:82]: delete this line

File: src/graph/graph_hub.h [hunk 7]
```diff
@@ -67,10 +67,17 @@ int DeleteEdge
...
+int UpdateEntityProperties
 (
 	GraphContext *gc,            // graph context to update the entity
 	GraphEntity *ge,             // the entity to be updated
 	const AttributeSet set,      // attributes to update
 	GraphEntityType entity_type  // the entity type (node/edge)
 );
+
+
+int UpdateNodeLabels(
```
Comment (AviAvni) [inline src/graph/graph_hub.h:79]: add documentation to this function

File: tests/flow/test_entity_update.py [hunk 8]
```diff
@@ -234,3 +234,129 @@ def test17_invalid_entity_assignment(self):
                 self.env.assertTrue(False)
             except ResponseError as e:
                 self.env.assertContains("Property values can only be of primitive types or arrays of primitive types", str(e))
+
+    def test18_update_node_label(self):
```
Comment (AviAvni) [inline tests/flow/test_entity_update.py:238]: add tests for 
```
MATCH (n) SET x:L
MATCH (n) SET x:L:L:L
```

File: tests/flow/test_entity_update.py [hunk 9]
```diff
@@ -234,3 +234,129 @@ def test17_invalid_entity_assignment(self):
...
+        result = graph.query("MATCH (n:TestLabel) RETURN n")
+        self.env.assertEqual(len(result.result_set), 0)
+        result = graph.query("MATCH (n) SET n:TestLabel")
+        self.env.assertEqual(result.labels_added, 1)
+        result = graph.query("MATCH (n:TestLabel) RETURN n")
+        self.env.assertEqual(len(result.result_set), 1)
```
Comment (AviAvni) [inline tests/flow/test_entity_update.py:244]: check also that we get the correct label in the results

File: src/execution_plan/ops/op_update_labels.c (hunk 2 above)
Comment (DvirDukhan) [reply src/execution_plan/ops/op_update_labels.c:122]: I guess this is GitHub way of showing things? 
image

File: src/ast/ast_shared.h (hunk 4 above)
Comment (DvirDukhan) [reply src/ast/ast_shared.h:57]: image

File: src/execution_plan/ops/op_update_labels.c (hunk 2 above)
Comment (AviAvni) [reply src/execution_plan/ops/op_update_labels.c:122]: look if all the indentation using \t and not space

File: src/ast/ast_build_op_contexts.c (hunk 3 above)
Comment (AviAvni) [reply src/ast/ast_build_op_contexts.c:142]: space vs tabs?

Comment (DvirDukhan) [reply src/ast/ast_build_op_contexts.c:142]: tabs before spaces after, right?

Comment (AviAvni) [reply src/ast/ast_build_op_contexts.c:142]: yes

**Review by AviAvni — CHANGES_REQUESTED** [review]
great work

**Review by AviAvni — APPROVED** [review]

**Pull request merged**